File: Data Structures & Algorithms/max-water-container/submission-3.kt
class Solution {
    fun base(i: Int, j: Int): Int = j - i
    fun height(heights: IntArray, i: Int, j: Int): Int = min(heights[i], heights[j])

    fun getCurrentArea(heights: IntArray, i: Int, j: Int): Int = base(i, j) * height(heights, i, j)

    fun maxArea(heights: IntArray): Int {
        if (heights.size < 2) throw IllegalArgumentException("Array size must be larger than 1")
        var left = 0
        var right = heights.size - 1
        var largestArea = 0

        while (left < right) {
            largestArea = max(getCurrentArea(heights, left, right), largestArea)
            val leftHeight = heights[left]
            val rightHeight = heights[right]
            when {
                leftHeight < rightHeight -> left++
                leftHeight > rightHeight -> right--
                else -> left++
            }
        }

        return largestArea
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-3.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left = 0
        right = len(nums) - 1
        minimum = nums[0]

        while left <= right:
            mid = (left + right) // 2
            if nums[left] < nums[right]:
                minimum = min(nums[left], minimum)
                break
            
            minimum = min(nums[mid], minimum)

            if nums[mid] >= nums[left]:
                left = mid + 1
            else:
                right = mid - 1

        return minimum